tcmalloc/internal/percpu_tcmalloc.h

```diff
@@ -357,13 +357,143 @@ static inline ABSL_ATTRIBUTE_ALWAYS_INLINE int TcmallocSlab_Push(
   int cpu = VirtualRseqCpuId();
   return f(cpu, cl, item);
 }
-#endif  // defined(__x86_64__)
+#endif // defined(__x86_64__)
+
+#if defined(__aarch64__)
+
+template <size_t Shift, size_t NumClasses>
+static inline ABSL_ATTRIBUTE_ALWAYS_INLINE int TcmallocSlab_Push(
+    typename TcmallocSlab<Shift, NumClasses>::Slabs* slabs, size_t cl,
+    void* item, OverflowHandler f) {
+  void* region_start;
+  uint64_t cpu_id;
+  void* slab_header;
+  uintptr_t current;
+  uintptr_t end;
+#if TCMALLOC_PERCPU_USE_RSEQ_ASM_GOTO
+  asm goto(
+#else
+  bool overflow;
+  asm volatile(
+#endif
+      // TODO(b/141629158):  __rseq_cs only needs to be writeable to allow for
+      // relocations, but could be read-only for non-PIE builds.
+      ".pushsection __rseq_cs, \"aw?\"\n"
+      ".balign 32\n"
+      ".local __rseq_cs_TcmallocSlab_Push_%=\n"
+      ".type __rseq_cs_TcmallocSlab_Push_%=,@object\n"
+      ".size __rseq_cs_TcmallocSlab_Push_%=,32\n"
+      "__rseq_cs_TcmallocSlab_Push_%=:\n"
+      ".long 0x0\n"
+      ".long 0x0\n"
+      ".quad 4f\n"
+      ".quad 5f - 4f\n"
+      ".quad 2f\n"
+      ".popsection\n"
+#if !defined(__clang_major__) || __clang_major__ >= 9
+      ".reloc 0, R_AARCH64_NONE, 1f\n"
+#endif
+      ".pushsection __rseq_cs_ptr_array, \"aw?\"\n"
+      "1:\n"
+      ".balign 8;"
+      ".quad __rseq_cs_TcmallocSlab_Push_%=\n"
+      // Force this section to be retained.  It is for debugging, but is
+      // otherwise not referenced.
+      ".popsection\n"
+      ".pushsection .text.unlikely, \"ax?\"\n"
+      ".long %c[rseq_sig]\n"
+      ".local TcmallocSlab_Push_trampoline_%=\n"
+      ".type TcmallocSlab_Push_trampoline_%=,@function\n"
+      "TcmallocSlab_Push_trampoline_%=:\n"
+      "2:\n"
+      "b 3f\n"
+      ".popsection\n"
+      // Prepare
+      //
+      // TODO(b/151503411):  Pending widespread availability of LLVM's asm
+      // goto with output contraints
+      // (https://github.com/llvm/llvm-project/commit/23c2a5ce33f0), we can
+      // return the register allocations to the compiler rather than using
+      // explicit clobbers.  Prior to this, blocks which use asm goto cannot
+      // also specify outputs.
+      "3:\n"
+      // Use current as scratch here to hold address of this function's
+      // critical section
+      "adrp %[current], __rseq_cs_TcmallocSlab_Push_%=\n"
+      "add  %[current], %[current], :lo12:__rseq_cs_TcmallocSlab_Push_%=\n"
+      "str %[current], [%[rseq_abi], %c[rseq_cs_offset]]\n"
+      // Start
+      "4:\n"
+      // cpu_id = __rseq_abi.cpu_id;
+      "ldr %w[cpu_id], [%[rseq_abi], %[rseq_cpu_offset]]\n"
+      // region_start = Start of cpu region
+      "add %[region_start], %[slabs], %[cpu_id], LSL %c[shift]\n"
+      // slab_header = slab header addr
+      "add %[slab_header], %[region_start], %[cl], LSL #3\n"
+      // current = slabs->current (current index)
+      "ldrh %w[current], [%[slab_header]]\n"
```

Comment:
Do these have to be 16-bit? If they could be 64-bit, you could fold one of the adds into the addressing mode.

Reply:
I liked your idea here. However making this a 64-bit load would involve having to change the pointers in the headers to 64-bit, which mean the total header (current, begin end and end_copy) would not fit in 64-bits and that would require changing the way we compute the class-index. I.e. we would no longer be able to multiply it by 8. This gave me the idea to hoist the multiplication by 8, replace the slab_header add with a end/begin offset add and fold the new `cl_lsl3` into the loads.